enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,191 @@
+---
+title: microshift-apiserver-custom-certs
+authors:
+  - "@eslutsky"
+reviewers:
+  - "@jerpeter, MicroShift contributor"
+  - "@pacevedom, MicroShift contributor"
+  - "@ggiguash, MicroShift contributor"
+  - "@benluddy, OpenShift API server team"
+  - "@standa, OpenShift Auth Team"
+
+approvers:
+  - "@jerpeter"
+api-approvers:
+  - N/A
+creation-date: 2021-01-18
+last-updated: 2023-01-18
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-2101
+see-also:
+  - microshift-apiserver-certs.md
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# MicroShift API server custom external certificates
+## Summary
+
+This enhancement extends the Microshift apiserver Certs to allow the user to
+configure the use of client certificates for api server external TLS access.
+
+
+> Anytime the document mentions API server, it refers to kube API server.
+
+
+## Motivation
+Currently, users of Microshift cannot provide their own generated certificates,
+Some customers have very strict requirements regarding TLS certs. There are frequently intermediate proxies which allow only TLS traffic with certs that are signed by organization owned CAs.
+
+### User Stories
+* As a MicroShift administrator, I want to be able to add  organization generated certificates.
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate.
+* As a MicroShift administrator, I want to be able to remove organization generated certificates.
+* As a MicroShift administrator, I want to be able to remove DNS names for each certificate.
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+N/A
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file.
+
+a new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+
+for each provided certificate, The following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
```

Comment:
From user stories:
> * As a MicroShift administrator, I want to provide additional DNS names for each certificate.
> * As a MicroShift administrator, I want to be able to remove DNS names for each certificate.

Just want to make sure: whatever is given in the `names` will completely overwrite names from the cert? So if someone wants to add something, they will need to include names already existing in the cert, and if they want to remove single name, they will need to provide everything but that name?

Reply:
no it wont overwrite,  it will be in additional to the names in the Cert, so both check will apply .
so its not possible to remove names from within the certs ..

Reply:
by `remove DNS names` i meant removing it from the `names` entries, maybe its misleading statement that we support removing DNS names ..

Reply:
Ah, you meant list in the config: `apiServer.namedCertificates[].names`?
If that's the case then I think user stories are a bit misleading.

Reply:
yeap . i'll remove it then